### DISCUSSION TITLE
Flexible channels mapping/configurations

### DISCUSSION THREAD

**Discussion by sdellava** [ExpressLRS/ExpressLRS#2604] [Ideas]
This idea stems from the desire to make radio control channel configuration management mapped to ELRS modes more flexible and at the same time more efficient.

Currently, we have 4 modes available that allow 12 to 16 channels to be transmitted using the 10 bytes of which the packet sent from the tx to the rx is composed.

As much as the various modes may be suitable to meet most needs, each mode has adverse effects that can significantly impact the 
usability of the ELRS system.

For example, the Hybrid mode has only 4 10-bit channels that can be used to control servos and 7 d 3-bit channels that are generally wasted since radios generally have only 3-position switches that can be handled with only 2 bits.

If you need more than 12 channels of which 6 are 10 bits you have to use the Full Res 16ch Rate mode which however is only available with 333 Hz packet rate and half rate transmission. 

**The solution I propose allows the user to configure completely discretionary channel mapping to the 79 available channels (10x8bit - 1 bit for arm which is always sent).**

For simplicity, I describe the solution from the perspective of the user who has to configure the mapping between channels and bits, then explain how to use the information introduced by the user.

The user is shown a table to associate the channels of the remote control with the available bits. The following table represents an example configuration where in addition to the four 10-bit channels, another nineteen 2-bit channels and one 1-bit channel are mapped, for a total of 24 channels (that's plus the ARM).

![image](https://github.com/ExpressLRS/ExpressLRS/assets/11772150/16aa2a59-78cc-4236-9fb6-c9ba4693dff7)

This mapping makes it possible to double the number of channels handled regardless of packet rate and without having to transmit in half duplex.

Of course, any combination is possible, going up to a maximum of seven 10-bit channels, with the remaining 9 bits configurable as desired. In example:

![image](https://github.com/ExpressLRS/ExpressLRS/assets/11772150/2758a1dd-31b5-4044-9cda-c3820d8c37b5)

What can be done, however, if the user needs ten 10-bit channels or otherwise a configuration that cannot be mapped to the only available 79 bits?

The solution is to present the client with a table consisting of 4 columns. Each column represents a sending batch, which I explain later.

Take for example:

![image](https://github.com/ExpressLRS/ExpressLRS/assets/11772150/3795a6a4-3e93-42d0-bcc1-abb352149e47)

As you can see, bits 1 to 61 are transmitted by mapping the same channels for each batch. This implies that there is no additional latency in the transmission of these channels. They are transmitted according to the packet rate requested by the user.

In contrast, bits 62 to 80 are mapped differently for each batch. This allows multiple channels to be mapped, and each channel will have an added latency that depends only on how many batches need to be sent before mapping it back into its assigned bits.

Of course, the rapresentation is simplified: the bits used for a 10-bit channel in batch 1 need not necessarily be used in subsequent batches for 10-bit channels as well. 

The advantage of this approach is thus the ability to freely define both the number of channels and which of them can be transmitted "slower"

Now that we have a simple and efficient way to allow the user to map channels visually (the web configurator can of course present pre-made configurations) let us see how the sw can handle the map entered by the user.

The idea is to create a table consisting of 4 columns and n rows, where n = 80 x number of batches 

Each row represents one bit of the packet being sent

In the first column the number of the channel a mapped is entered.
In the second column the number of bits assigned to the channel.
In the third column the bit id is entered (e.g., the second bit of a three-bit channel will have the value 1 here) 
In the fourth column the value of the bit (1 or zero) is entered.

A tread reads each row of the table, converts the channel value received from the remote control to the number of bits given in the second column, and values the last column of the next n rows.

A second tread prepares the bit packet to be transmitted by simply reading the n 80 bits the fourth column. For the first batch it uses bits 1 to 80, for the second 81 to 160, for the third 161 to 240.

To indicate to the receiver which batch the packet it received refers to, we indicate its number in the header. If it should not be possible to indicate the batch number, using 2 bits, then the last two bits of the 80 of each batch can be used as the identification number. 

The receiver receives the channel mapping configuration as soon as the link is established or possibly, for added security, at binding. In this way, there is no need to resend  channel mapping in case the link is lost.

The receiver reverse interprets the received data and passes it to the fc.

**Comment by sdellava** [comment]
UPDATE: to simplfy the user experience the configurator can propose some pre-defined channel mapping described in detail in the expressrls.org web sited. To support custom confgurations a cfg file can computed offline and uploaded in the configurator.

**Comment by pkendall64** [comment]
For 2.4GHz, the only 10 byte payload modes are 333Hz Full and 100Hz Full. All the other modes ave a 5 byte payload.
For 900, 100Hz full is the only 10 byte payload mode, all the others are 5 byte payload.

So this proposal would only work for 333 or 100 Hz modes and these modes already allow selecting 8, 12 or 16 channels of full resolution data.

EDIT:  And CRSF only allows 16 channels of data.

**Comment by sdellava** [comment]
Thank you @pkendall64 for your feedback.

Of course, the method I presented is independent of the size of the transmitted packet and even to the number of sending batches.

The point of my proposal is to give the user the ability to decide how to configure his link with a greater degree of freedom, specifically deciding if how it should be updated.

**In fact, this method is an extension of the switch mode hybrid, which involves sending a packet containing the values of seven AUXs in rotation.**

Here is an example of a 5-byte mapping (plus the ARM bit) with only four batches, that maps 16 channels (arm, four 10-bits, eight 3-bits and five 2-bits):

![image](https://github.com/ExpressLRS/ExpressLRS/assets/11772150/d86c7606-3ee5-4ca2-bbee-a2fde2b928aa)

** CH24 is a mistake

Obviously, to transmit more than 16 channels, a different communication protocol must be implemented between remote control and radio.

We could greatly simplify the ELRS system if this protocol were designed ad hoc in collaboration with the EDGTX community, but even without considering this assumption, there are several options that support 32 channels:
- SRXL2 multiplex
- Jeti ExBus
- HoTT (supports up to 32 channels and also includes advanced telemetry capabilities)

In a longer-term view, CAN BUS support could also be assumed, but this is off topic.

**Reply by sdellava** [reply]
Update: just in case someone want to experiment with HoTT protocol, here you can find some free code: 
https://github.com/Stapelzeiger/HoTT
